### PR TITLE
feat(stream): support stream null safe hash join

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Support stream null safe hash join.
- Remove `JoinHashKey`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#4958 #5097

### CODE REVIEW

File: src/common/src/hash/key.rs [hunk 1]
```diff
@@ -164,16 +165,17 @@ pub trait HashKey:
 /// | (1, null) | (1, null) | true |
 /// | (null, 1) | (null, 1) | false |
 /// | (null, null) | (null, null) | false |
-pub struct JoinHashKey<'a, K> {
+#[derive(Clone)]
+pub struct JoinHashKey<K> {
     key: K,
-    null_matched: &'a FixedBitSet,
+    null_matched: Arc<FixedBitSet>,
```
Comment (TennyZhuang) [inline src/common/src/hash/key.rs:171]: In most scenarios, the bitset is smaller than 1 byte, so just clone it directly without Arc is more efficient.

Comment (chenzl25) [reply src/common/src/hash/key.rs:171]: Make sense.

Comment (fuyufjh) [inline src/common/src/hash/key.rs:171]: Is this `null_matched` identical for all the entries in a hash table? If so, can we eliminate it from the key?

Comment (chenzl25) [reply src/common/src/hash/key.rs:171]: Yes, they are identical. Currently, `null_matched` is mostly used to overwrite `PartialEq` trait.
```rust
impl<K: HashKey> PartialEq for JoinHashKey<K> {
    fn eq(&self, other: &Self) -> bool {
        self.key == other.key && self.key.null_bitmap().is_subset(&self.null_matched)
    }
}
```

Comment (yuhao-su) [reply src/common/src/hash/key.rs:171]: One possible alternative way is to make the `self.key.null_bitmap().is_subset(&self.null_matched)` to the join executor. If it's false, then no matched data in a hash map.

Comment (chenzl25) [reply src/common/src/hash/key.rs:171]: Agree. Let me do some refactor.

Comment (chenzl25) [reply src/common/src/hash/key.rs]: Please take a look again @yuhao-su and I also remove `JoinHashKey` from `BatchHashJoinExecutor`

**Comment by yuhao-su** [comment]
LGTM!

**Review by yuhao-su — APPROVED** [review]

**Pull request merged**